nixos/firefox: make variant librewolf

mBornand opened the pull request:
Firefox has several patched versions, some of them like librewolf use different configuration paths, but can be configured in a similar way. Additionally we may want to have both variant arround.

- add a function to create the common base.
- add a module for librewolf

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [x] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mBornand committed:
nixos/firefox: make variant librewolf

Firefox has several patched versions, some of them like librewolf use
different configuration paths, but can be configured in a similar way.
Additionally we may want to have both variant arround.

- add a function to create the common base
- add a module for librewolf
- add module argument to tests

maevii reviewed (commented):
Just FYI, I did something similar recently for K3s/RKE2 in #447847. Not sure whether how I structured the modules is better or not, just something to be aware of.

mBornand commented on nixos/modules/programs/firefox/librewolf.nix:
```diff
@@ -0,0 +1,21 @@
+    (firefoxLib.mkFirefoxBaseModule {
+      variant = "librewolf";
+      prettyName = "Librewolf";
+      defaultPackage = pkgs.librewolf;
```
I was hesitant because of #416711

azey7f replied:
I think for now we should still put this in there until that PR gets merged, right now it's still a separate package.

mBornand commented:
@azey7f I actually prefer the way you did in #447847, so I made something similar.

azey7f commented on nixos/modules/module-list.nix:
```diff
@@ -210,7 +210,7 @@
   ./programs/fcast-receiver.nix
   ./programs/feedbackd.nix
-  ./programs/firefox.nix
+  ./programs/firefox
```
```suggestion
  ./programs/firefox/default.nix
```
AFAIK `default.nix` is usually specified explicitly even though it isn't technically needed

azey7f commented on nixos/modules/programs/firefox/default.nix:
```diff
@@ -0,0 +1,308 @@
+            extraPrefsFiles =
+              (old.extraPrefsFiles or [ ])
+              ++ cfg.autoConfigFiles
+              ++ [ (pkgs.writeText "autoconfig.js" cfg.autoConfig) ];
```
```suggestion
              ++ [ (pkgs.writeText "${variant}-autoconfig.js" cfg.autoConfig) ];
```
I still feel like this should include the name, with `firefox-autoconfig.js` it's easy to tell at a glance what the file is while just `autoconfig.js` could be anything.

azey7f commented on nixos/modules/programs/firefox/lib.nix:
```diff
@@ -0,0 +1,302 @@
+            extraPrefsFiles =
+              (old.extraPrefsFiles or [ ])
+              ++ cfg.autoConfigFiles
+              ++ [ (pkgs.writeText "firefox-autoconfig.js" cfg.autoConfig) ];
```
I still feel like it should be there, with `firefox-autoconfig.js` it's easy to tell at a glance what the file is (e.g. just browsing through /nix/store or a cache) while just `autoconfig.js` could be anything.

azey7f commented on nixos/tests/firefox.nix:
```diff
@@ -18,7 +19,7 @@
       environment.systemPackages = [ pkgs.xdotool ];
 
-      programs.firefox = {
+      programs."${firefoxModule}" = {
```
```suggestion
      programs.${firefoxModule} = {
```

maevii reviewed (changes requested):
One last thing, then it should be all from me. You can also resolve review comments to keep track of them better, non-committer reviewers like me can't for some reason.

maevii reviewed (approved):
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 467398 --additional-package nixosTests.firefox-esr-140 --additional-package nixosTests.firefox-beta --additional-package nixosTests.firefox-devedition --additional-package nixosTests.firefox --additional-package nixosTests.librewolf --additional-package nixosTests.firefox-esr`
Commit: `59fee27797a7ddf7a69681a9f49bea793f658c3f`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 5 tests built:</summary>
  <ul>
    <li>nixosTests.firefox</li>
    <li>nixosTests.firefox-beta</li>
    <li>nixosTests.firefox-devedition</li>
    <li>nixosTests.firefox-esr (nixosTests.firefox-esr-140)</li>
    <li>nixosTests.librewolf</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/6761

Hythera commented:
I don't think we can inherit the language packs option from Firefox as Librewolf requires their language packs to be patched with their additional translations. They created a patch so the official language packs will automatically be removed due to them causing issues with the permission system. Also see #509190 for more context.

mBornand committed:
nixos/firefox: make languagepack exclusive to firefox

mBornand committed:
nixos/firefox: make languagepack exclusive to firefox

mBornand commented:
> I don't think we can inherit the language packs option from Firefox as Librewolf requires their language packs to be patched with their additional translations. They created a patch so the official language packs will automatically be removed due to them causing issues with the permission system. Also see #509190 for more context.

Done, now we can only set the languagepack for firefox.

This also provids a way of showing how to add options to one variant only.

mBornand committed:
nixos/firefox: make variant librewolf

Firefox has several patched versions, some of them like librewolf use
different configuration paths, but can be configured in a similar way.
Additionally we may want to have both variant arround.

- add a function to create the common base
- add a module for librewolf
- add module argument to tests

Hythera reviewed (approved):
Don't have much of on a opinion on this. Tested both `firefox` and `librewolf` modules. No issues. Just make sure to resolve the merge conflicts (again).

danth commented on nixos/modules/programs/firefox/default.nix:
```diff
@@ -0,0 +1,194 @@
+  ];
+
+  meta.maintainers = with lib.maintainers; [
+    danth
```
As of c27a1c00a6a4698f168cc06c749f2091a3b5f77e I am no longer a maintainer for this.

mBornand replied:
Sorry missed this one in a previous rebase, fixed.

ismay-tue commented:
I've been looking forward to this landing in master. Is this just waiting on the CI checks, or are there still issues that need to be addressed?
